File: it-agents/skills/code-architect/references/code-examples/ruby/flyweight.rb
# frozen_string_literal: true

# Flyweight — share intrinsic state across many objects to save memory.

# Flyweight holds only intrinsic (shared, immutable) state.
class GlyphStyle
  attr_reader :font, :size, :color

  def initialize(font, size, color)
    @font = font
    @size = size
    @color = color
  end

  # Extrinsic state (position) is passed in at use time, not stored.
  def stamp(char, at:)
    "'#{char}'@#{at} {#{@font}/#{@size}/#{@color}}"
  end
end

# Factory caches and reuses flyweights by their intrinsic key.
class StyleFactory
  def initialize = @pool = {}

  def fetch(font, size, color)
    key = [font, size, color]
    @pool[key] ||= GlyphStyle.new(font, size, color)
  end

  def distinct_count = @pool.size
end

if __FILE__ == $PROGRAM_NAME
  factory = StyleFactory.new
  text = "aaab"

  text.chars.each_with_index do |char, i|
    style = factory.fetch("mono", 12, "black")
    puts style.stamp(char, at: i)
  end

  puts "characters drawn: #{text.length}"
  puts "shared flyweights: #{factory.distinct_count}"
end

File: it-agents/skills/code-architect/references/code-examples/typescript/Flyweight.ts
// Flyweight — share intrinsic state across many objects to save memory.

// Intrinsic, shareable state.
class GlyphStyle {
  constructor(
    public readonly font: string,
    public readonly size: number,
  ) {}

  describe(): string {
    return `${this.font}@${this.size}`;
  }
}

// Factory returns shared instances keyed by intrinsic state.
class GlyphStyleFactory {
  private pool = new Map<string, GlyphStyle>();

  get(font: string, size: number): GlyphStyle {
    const key = `${font}:${size}`;
    let style = this.pool.get(key);
    if (!style) {
      style = new GlyphStyle(font, size);
      this.pool.set(key, style);
    }
    return style;
  }

  get count(): number {
    return this.pool.size;
  }
}

// Extrinsic state (position) stays outside the flyweight.
interface PlacedGlyph {
  char: string;
  x: number;
  style: GlyphStyle;
}

function demo(): void {
  const factory = new GlyphStyleFactory();
  const text = "hello";
  const glyphs: PlacedGlyph[] = [...text].map((char, x) => ({
    char,
    x,
    style: factory.get("Inter", 12),
  }));
  console.log(`${glyphs.length} glyphs share ${factory.count} style object(s)`);
}

demo();
